### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by samira-barouti**
This refactoring PR aims to leverage OOP principles to implement Policy and PolicyEngine.
`Policy` is an interface that has four methods:
* GetName() which returns the name of policy
* GetMetadata() which returns the metadata related to the policy
* GetHelp() which returns the help text
* Validate() which takes an image tag and validates whether the image satisfies certain criteria

Each policy is a type (i.e construct) that must implement above functions. 

The PolicyRunner takes a slice of Policy types and run the Validate() method on each policy with the given Image ID and aggregates the result.

### CODE REVIEW

**Comment by skattoju** [comment]
@samira-barouti potentially controversial but apparently the convention is to not prefix getters with Get : https://golang.org/doc/effective_go#Getters

**Comment by komish** [comment]
Thanks @samira-barouti! I'm reviewing shortly, and will add some comments in line. I'll also add the notes of our discussion after I review so we can loop in others to our discussion, and have them make a decision on how we proceed :+1:

First thing, can you `--amend` your commit message "Refactoring" to include more description for the commit log?

File: certification/formatters/formatters.go [hunk 1]
```diff
@@ -1,95 +1,89 @@
...
+
+// 		return nil, e
+// 	}
+
+// 	return responseJSON, nil
+// }
+
+// func JUnitXMLFormatter(r runtime.Results) ([]byte, error) {
+// 	return nil, fmt.Errorf("%w: The JUnit XML Formatter is not implemented", errors.ErrFeatureNotImplemented)
+// }
```
Comment (bcrochet) [inline certification/formatters/formatters.go:89]: Just remove the code. We don't want to commit large blocks of commented code.

File: certification/internal/policy/base_on_ubi.go [hunk 2]
```diff
@@ -36,14 +31,22 @@ var basedOnUBIValidatorFunc = func(image string) (bool, error) {
 	return false, nil
 }
 
-var basedOnUBIPolicyMeta = Metadata{
-	Description:      "Checking if the container's base image is based on UBI",
-	Level:            "best",
-	KnowledgeBaseURL: "https://connect.redhat.com/zones/containers/container-certification-policy-guide", // Placeholder
-	PolicyURL:        "https://connect.redhat.com/zones/containers/container-certification-policy-guide",
+func (p BasedOnUbiPolicy) GetName() string {
```
Comment (bcrochet) [inline certification/internal/policy/base_on_ubi.go:34]: I will concur with Sid. Don't use 'Get'

**Comment by komish** [comment]
OK, as I discussed with @samira-barouti -I'd like to get at least 1 additional reviewer here to chime in as this PR changes a few things. The repo is largely lacking design docs at this moment (I'll get it done here shortly), so here's the summary as I see it (@samira-barouti to correct anything that I misunderstand, please!):

---

Previously, each policy leveraged a generic implementation (called `Definition`) that implemented the Policy interface. This was an easy way to bind a valid function (called `ValidatorFunc`) with some additional metadata into a single struct that could be used to implement the Policy interface. I have a another PR pending review that does similar for formatters. In the case of the `Definition`, this is not publicly available to external use, but I think it would have been beneficial to provide that.

In this PR, the Policies are instead bound to unique structs that each implement the Policy interface independently. The ability to build your own policy implementations was 100% part of the original design goal, as we're leaving this open ended so that (a) developers can implement their own policies if they would like to do so, and (b) so that we can continue to evolve how we design/enforce our policies, should the implementations change (E.g. current policy enforcement in PoC repos are done using cmd.Exec to Podman and similar tools, but that could change in the future and we want to support that). 

---
The pros of this approach are that you can bind policy-specific data to their struct (which is the major driver I could see to use your own policy struct in general). 

The con here is that, at least given the current cmd.Exec workflow, is that the current structs aren't using per-policy data. We're also not preserving the generic definition, so users would simply need to implement the Policy-related interfaces. I think adding in a generic definition is useful to get folks started, even if we don't use it ourselves.

There are some other changes as well, so whoever picks this up for additional review, please check out that discussion and let us know thoughts!

File: cmd/root.go [hunk 3]
```diff
@@ -37,14 +35,14 @@ var rootCmd = &cobra.Command{
 		}
 
 		runner.ExecutePolicies()
-		results := runner.GetResults()
+		// results := runner.GetResults()
```
Comment (bcrochet) [inline cmd/root.go:38]: If it isn't needed, remove it.

**Review by bcrochet — COMMENTED** [review]
Just a couple of comments about comments.

File: certification/formatters/formatters.go (hunk 1 above)
Comment (komish) [reply certification/formatters/formatters.go:89]: @bcrochet Actually, we need to go the other way. All of the output handlers were removed in this PR, and we need to rewire those in.

File: certification/internal/policy/less_than_max_layers.go [hunk 4]
```diff
@@ -1,27 +1,28 @@
...
 
-func UnderLayerMax() *Definition {
-	return &Definition{
-		ValidatorFunc: isUnderLayerMaxValidatorFunc,
-		Metadata:      isUnderLayerMaxPolicyMeta,
-		HelpText:      isUnderLayerMaxPolicyHelp,
-	}
+func (p UnderLayerMaxPolicy) Validate(image string) (bool, error) {
+	return true, nil
 }
```
Comment (komish) [inline certification/internal/policy/less_than_max_layers.go:8]: While these are all unimplemented, can these all return `false` and the Err indicating they're not implemented?

File: certification/internal/policy/types.go [hunk 5]
```diff
@@ -1,38 +1,27 @@
...
+// to use and identify a given policy.
+type Policy interface {
+	// Validate whether the asset enforces the policy.
+	Validate(image string) (result bool, err error)
+	// return the name of the policy
+	GetName() string
+	// return the policy's metadata
+	GetMetadata() Metadata
+	// return the policy's help text
+	GetHelp() HelpText
```
Comment (komish) [inline certification/internal/policy/types.go:13]: Defining the policy here means that it is no longer an exported API. This needs to live in `certification.go`, or somewhere outside of the `internal/` subdirectory.

File: certification/internal/policy/types.go [hunk 6]
```diff
@@ -1,38 +1,27 @@
...
 	Level            string `json:"level" xml:"level"`
 	KnowledgeBaseURL string `json:"knowledge_base_url,omitempty" xml:"knowledgeBaseURL"`
 	PolicyURL        string `json:"policy_url,omitempty" xml:"policyURL"`
 }
 
-// HelpText is the help message associated with any given policy.
+// HelpText is the help message associated with any given policy
 type HelpText struct {
 	Message    string `json:"message" xml:"message"`
 	Suggestion string `json:"suggestion" xml:"suggestion"`
```
Comment (komish) [inline certification/internal/policy/types.go:27]: Nitpick - I believe the punctuation here is 👍  from godoc's perspective. I've been inconsistent in other places where I've missed punctuation (my apologies). Probably better to leave this in place, and maybe add it in where it's missing. You don't have to add other missing periods, but we can leave this one here.

File: cmd/root.go [hunk 7]
```diff
@@ -37,14 +35,14 @@ var rootCmd = &cobra.Command{
...
-		if err != nil {
-			return err
-		}
+		// // formattedResults, err := formatters.GenericJSONFormatter(results)
+		// // if err != nil {
+		// // 	return err
+		// // }
 
-		fmt.Fprint(os.Stdout, string(formattedResults))
+		// // fmt.Fprint(os.Stdout, string(formattedResults))
```
Comment (komish) [inline cmd/root.go:45]: We need to wire the output formatting back in. There's a separate PR that actually makes it dynamic based on user configuration, so returning it back to the state it was in should work.

**Review by komish — CHANGES_REQUESTED** [review]
Just a few changes here

File: certification/formatters/formatters.go (hunk 1 above)
Comment (bcrochet) [reply certification/formatters/formatters.go:89]: That's fine. Just don't want the final commit to contain large blocks of commented code.

File: certification/internal/policy/types.go (hunk 5 above)
Comment (bcrochet) [reply certification/internal/policy/types.go:13]: A top-level policy package would be where I would put it.

**Pull request closed without merging**